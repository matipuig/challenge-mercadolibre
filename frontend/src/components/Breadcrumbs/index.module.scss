@import '~/scss/variables/colors';

.breadcrumbsContainer {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;
  padding: 10px;
}

.breadcrumb {
  color: $secondaryColor;
  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
}
