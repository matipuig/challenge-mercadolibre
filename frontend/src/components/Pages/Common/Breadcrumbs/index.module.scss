@use '~/scss/variables/colors' as colors;

.breadcrumbsContainer {
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  gap: 10px;
  padding: 10px;
}

.breadcrumb {
  color: colors.$secondary;
  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
}
