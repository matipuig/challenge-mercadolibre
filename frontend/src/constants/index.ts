/**
 * Facade to access constants.
 */

import LOGS from './logs';
import ROUTES from './routes';
import SERVICES from './services';

const AUTHOR = 'Mat√≠as Puig';

export const CONSTANTS = { AUTHOR, LOGS, ROUTES, SERVICES };

export default CONSTANTS;
