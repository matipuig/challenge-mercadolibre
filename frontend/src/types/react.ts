/**
 * Contains the different types necessary to use React with typescript.
 */
import React, { PropsWithChildren, ReactElement as Element } from 'react';

export type ComponentChildren = React.ReactNode;
export type ReactElement = Element;
export type FunctionalComponent<P = unknown> = (props: P) => ReactElement | null;
export type FunctionalComponentWithChildren<P = unknown> = FunctionalComponent<
  PropsWithChildren<P>
>;
