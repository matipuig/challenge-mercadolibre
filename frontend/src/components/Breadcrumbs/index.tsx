/**
 * Contains the breadcrumbs component.
 */
import { Fragment, ReactElement } from 'react';
import Link from 'next/link';

import CONSTANTS from '~/constants';
import { ItemCategory } from '~/types/services/backend';
import styles from './index.module.scss';

interface BreadcrumbProps {
  categories: ItemCategory[];
}

const { ROUTES } = CONSTANTS;

const getDestinationRoute = (categoryId: string): string => {
  const queryParams = new URLSearchParams();
  queryParams.set('category', categoryId);
  return `${ROUTES.MAIN}?${queryParams}`;
};

export const Breadcrumbs = ({ categories }: BreadcrumbProps): ReactElement => {
  const lastCategoryIndex = categories.length - 1;
  return (
    <div className={styles.breadcrumbsContainer}>
      {categories.map((e, i) => (
        <Fragment key={e.id}>
          <Link href={getDestinationRoute(e.id)}>
            <a className={styles.breadcrumb}>{e.name}</a>
          </Link>
          {i < lastCategoryIndex && <div className={styles.separator}>&gt;</div>}
        </Fragment>
      ))}
    </div>
  );
};

export default Breadcrumbs;
