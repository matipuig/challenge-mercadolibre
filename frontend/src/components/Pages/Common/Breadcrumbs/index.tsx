/**
 * Contains the breadcrumbs component.
 */
import { Fragment, ReactElement } from 'react';

import { t } from 'i18next';
import Link from 'next/link';

import { CONSTANTS } from '~/constants';
import { getAvailableI18nTexts } from '~/i18n';
import { ItemCategory } from '~/types/services/backend';

import styles from './index.module.scss';

interface BreadcrumbProps {
  categories: ItemCategory[];
}

const { ROUTES } = CONSTANTS;
const texts = getAvailableI18nTexts();
const { ariaLabel } = texts.components.common.breadcrumbs;

const getDestinationRoute = (categoryId: string): string => {
  const queryParams = new URLSearchParams();
  queryParams.set('category', categoryId);
  return `${ROUTES.ITEMS}?${queryParams}`;
};

export const Breadcrumbs = ({ categories }: BreadcrumbProps): ReactElement => {
  const lastCategoryIndex = categories.length - 1;
  return (
    <nav className={styles.breadcrumbsContainer}>
      {categories.map((e, i) => (
        <Fragment key={e.id}>
          <Link href={getDestinationRoute(e.id)} aria-label={t(ariaLabel)}>
            <a className={styles.breadcrumb} data-testid="breadcrumbLink">
              {e.name}
            </a>
          </Link>
          {i < lastCategoryIndex && <div className={styles.separator}>&gt;</div>}
        </Fragment>
      ))}
    </nav>
  );
};

export default Breadcrumbs;
