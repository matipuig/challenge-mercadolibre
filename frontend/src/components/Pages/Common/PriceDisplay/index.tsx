/**
 * Contains the no results screen.
 */
import { ReactElement } from 'react';

import { useTranslation } from 'react-i18next';

import { getAvailableI18nTexts } from '~/i18n';
import { Price } from '~/types/services/backend';
import { makeHumanFriendly } from '~/utils/conversors';

import styles from './index.module.scss';

interface PriceDisplayProps {
  price: Price;
  withDecimals?: boolean;
}

const texts = getAvailableI18nTexts();
const { withPesos } = texts.components.common.priceDisplay;

const getDecimals = (decimals: number): string => {
  const strDecimals = decimals.toString();
  return strDecimals.length < 2 ? `0${strDecimals}` : strDecimals;
};

export const PriceDisplay = ({ price, withDecimals }: PriceDisplayProps): ReactElement => {
  const { t } = useTranslation();
  const { amount, decimals, currency } = price;
  const friendlyAmount = makeHumanFriendly(amount);
  const isPesos = currency === 'ARS';
  const currencyDisplay = isPesos ? '$' : currency;
  const ariaLabel = isPesos ? t(withPesos).replace('[?]', `${amount}`) : `${amount} ${currency}`;
  const finalDecimals = getDecimals(decimals);
  return (
    <div className={styles.container} aria-label={ariaLabel}>
      <span className={styles.amount}>
        {currencyDisplay} {friendlyAmount}
      </span>
      {withDecimals && (
        <span aria-hidden className={styles.decimals}>
          <sup>{finalDecimals}</sup>
        </span>
      )}
    </div>
  );
};

export default PriceDisplay;
