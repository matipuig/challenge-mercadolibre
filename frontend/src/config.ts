/**
 * Has the config for the application. Mainly loaded from the environment.
 * IMPORTANT: If you want NEXT to allow a variable to be visible in the client, it
 * MUST start with "NEXT_PUBLIC_". Otherwise it would be undefined.
 * NEXT_PUBLIC_ variables are not available using the function getFromEnvironmentOrError.
 */

import { LogLevels } from './types/logger';

const LOGS_LEVEL_LABEL = 'LOGS_LEVEL';

const getFromEnvironmentOrError = (variableName: string): string => {
  const value = process.env[variableName];
  if (typeof value === 'undefined') {
    throw new Error(`Environment variable "${variableName}" is not defined.`);
  }
  return value;
};

const getLogLevelFromEnvironment = (): LogLevels => {
  const value = process.env[LOGS_LEVEL_LABEL];
  if (typeof value === 'undefined') {
    return LogLevels.silly;
  }
  const validLogLevels: string[] = Object.values(LogLevels);
  if (!validLogLevels.includes(value)) {
    return LogLevels.silly;
  }
  return value as LogLevels;
};

const throwMissingEnvironmentVariablError = (environmentVariable: string): never => {
  throw new Error(`The environment variable "${environmentVariable}" has not been set.`);
};

export const CONFIG = {
  /**
   * Node environment.
   */
  NODE_ENV: getFromEnvironmentOrError('NODE_ENV'),

  /**
   * Application name.
   */
  APP_NAME: getFromEnvironmentOrError('npm_package_name'),

  /**
   * Application version.
   */
  APP_VERSION: getFromEnvironmentOrError('npm_package_version'),

  /**
   * Contains the base URL.
   */
  BASE_URL:
    process.env.NEXT_PUBLIC_FRONTEND_BASE_URL ||
    throwMissingEnvironmentVariablError('NEXT_PUBLIC_FRONTEND_BASE_URL'),

  /**
   * Sets logs level.
   */
  LOGS_LEVEL: getLogLevelFromEnvironment(),

  /**
   * Services configuration.
   */
  SERVICES: {
    /**
     * Own MercadoLibre API backend.
     */
    BACKEND: {
      /**
       * Base URL of the services.
       */
      BASE_URL: getFromEnvironmentOrError('BACKEND_BASE_URL'),

      /**
       * API Key to consume the backend.
       */
      API_KEY: getFromEnvironmentOrError('BACKEND_API_KEY'),
    },
  },
};

export default CONFIG;
