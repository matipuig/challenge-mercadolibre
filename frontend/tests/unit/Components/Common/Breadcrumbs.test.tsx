/**
 * Contains the unit tests for BreadCrumbs.
 */
import { cleanup, render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { CONSTANTS } from '../../../../src/constants';
import { mockedCategories } from '../../__mocks__/categories';
import { Breadcrumbs } from '../../../../src/components/Pages/Common/Breadcrumbs';
import { getURLWithBase } from '../../../utils';

const { ROUTES } = CONSTANTS;
const { QUERY_PARAMS } = ROUTES;

describe('Common breadcrumbs component', () => {
  let breadcrumbs: RenderResult;

  beforeEach(() => {
    breadcrumbs = render(<Breadcrumbs categories={mockedCategories} />);
  });

  afterEach(() => {
    cleanup();
  });

  it('has at least one testing category.', () => {
    expect(mockedCategories.length).toBeGreaterThan(0);
  });

  it('renders.', () => {
    expect(breadcrumbs.baseElement).toBeInTheDocument();
  });

  it('matches snapshot.', () => {
    expect(breadcrumbs).toMatchSnapshot();
  });

  it('has the nav tag.', () => {
    const navTag = breadcrumbs.getByRole('navigation');
    expect(navTag).toBeDefined();
  });

  it('has a link for each category.', () => {
    const links = breadcrumbs.getAllByTestId('breadcrumbLink');
    expect(links.length).toEqual(mockedCategories.length);
  });

  it('has the correct texts in links.', () => {
    const links = breadcrumbs.queryAllByTestId('breadcrumbLink');
    links.forEach((link, index) => {
      const expectedText = mockedCategories[index].name;
      expect(link.innerHTML).toEqual(expectedText);
    });
  });

  it('has the correct link.', () => {
    const links = breadcrumbs.queryAllByTestId('breadcrumbLink');
    links.forEach((link) => {
      const href = link.getAttribute('href');
      expect(href).toBeDefined();
      const entireUrl = getURLWithBase(href as string);
      const url = new URL(entireUrl);
      const urlParams = url.searchParams;
      expect(url.pathname).toEqual(ROUTES.ITEMS);
      expect(urlParams.get(QUERY_PARAMS.CATEGORY)).toBeDefined();
    });
  });
});
