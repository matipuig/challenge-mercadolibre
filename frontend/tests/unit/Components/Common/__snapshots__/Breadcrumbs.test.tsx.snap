// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common breadcrumbs component matches snapshot. 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <nav
        data-testid="breadcrumbs"
      >
        <a
          aria-label="Categoría Animales y Mascotas"
          data-testid="breadcrumbLink"
          href="/items?category=MLA1071&description=Animales-y-Mascotas"
        >
          Animales y Mascotas
        </a>
        <div
          aria-hidden="true"
        >
          &gt;
        </div>
        <a
          aria-label="Categoría Perros"
          data-testid="breadcrumbLink"
          href="/items?category=MLA1072&description=Perros"
        >
          Perros
        </a>
        <div
          aria-hidden="true"
        >
          &gt;
        </div>
        <a
          aria-label="Categoría Perros de Raza"
          data-testid="breadcrumbLink"
          href="/items?category=MLA1073&description=Perros-de-Raza"
        >
          Perros de Raza
        </a>
      </nav>
    </div>
  </body>,
  "container": <div>
    <nav
      data-testid="breadcrumbs"
    >
      <a
        aria-label="Categoría Animales y Mascotas"
        data-testid="breadcrumbLink"
        href="/items?category=MLA1071&description=Animales-y-Mascotas"
      >
        Animales y Mascotas
      </a>
      <div
        aria-hidden="true"
      >
        &gt;
      </div>
      <a
        aria-label="Categoría Perros"
        data-testid="breadcrumbLink"
        href="/items?category=MLA1072&description=Perros"
      >
        Perros
      </a>
      <div
        aria-hidden="true"
      >
        &gt;
      </div>
      <a
        aria-label="Categoría Perros de Raza"
        data-testid="breadcrumbLink"
        href="/items?category=MLA1073&description=Perros-de-Raza"
      >
        Perros de Raza
      </a>
    </nav>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
